### PR TITLE
PS-6003 : Re-enable assertions in os_event_global_destroy() (Post 8.0…

### PR DESCRIPTION
**Pull request by satya-bodapati**
….17-merge fix)

Problem:
--------
The total number of events at shutdown are not zero. The non-zero count comes from
mutexes.

Fix:
----
1. Revert ib_unintialized_mutex
2. os_event that are created as part of mutexes are now created as part of init()
   (Basically follow upstream logic)

### CODE REVIEW

File: storage/innobase/include/ib0mutex.h [hunk 1]
```diff
@@ -518,6 +521,10 @@ struct TTASEventMutex {
   void destroy() UNIV_NOTHROW {
     ut_ad(!m_lock_word.load(std::memory_order_relaxed));
 
+    /* We have to free the event before InnoDB shuts down. */
+    os_event_destroy(m_event);
+    m_event = 0;
```
Comment (percona-ysorokin) [inline storage/innobase/include/ib0mutex.h:526]:
```suggestion
    m_event = nullptr;
```

File: storage/innobase/include/sync0rw.h [hunk 2]
```diff
@@ -634,30 +634,18 @@ struct rw_lock_t
...
-  rw_lock_t() noexcept { ut_d(magic_n = RW_LOCK_MAGIC_N;) }
-
-  /** Copy constructor */
-  rw_lock_t(const rw_lock_t &other) noexcept : rw_lock_t() {
-    recursive = other.recursive;
-    cfile_name = other.cfile_name;
-    cline = other.cline;
-    is_block_lock = other.is_block_lock;
-  }
+  rw_lock_t() { magic_n = RW_LOCK_MAGIC_N; }
```
Comment (percona-ysorokin) [inline storage/innobase/include/sync0rw.h:639]:
```suggestion
  rw_lock_t() noexcept : magic_n{RW_LOCK_MAGIC_N} {}
```

File: storage/innobase/include/sync0rw.h [hunk 3]
```diff
@@ -634,30 +634,18 @@ struct rw_lock_t
 #ifdef UNIV_DEBUG
 /** Value of rw_lock_t::magic_n */
 #define RW_LOCK_MAGIC_N 22643
-#endif /* UNIV_DEBUG */
 
   /** Constructor */
-  rw_lock_t() noexcept { ut_d(magic_n = RW_LOCK_MAGIC_N;) }
-
-  /** Copy constructor */
-  rw_lock_t(const rw_lock_t &other) noexcept : rw_lock_t() {
```
Comment (percona-ysorokin) [inline storage/innobase/include/sync0rw.h:643]: I would also made copy constructor `=default` or even `=delete` if not needed.

File: storage/innobase/include/sync0rw.ic [hunk 4]
```diff
@@ -584,7 +584,7 @@ void pfs_rw_lock_create_func(
     const char *cfile_name,  /*!< in: file name where created */
     ulint cline)             /*!< in: file line where created */
 {
-  new (lock) rw_lock_t();
+  ut_d(new (lock) rw_lock_t());
```
Comment (percona-ysorokin) [inline storage/innobase/include/sync0rw.ic:587]: Placement new (and, therefore, constructor) is called only in debug mode. Is this intentional?
If yes, shouldn't there be a destructor call counterpart in the `_destroy` function?

**Comment by satya-bodapati** [comment]
https://ps80.cd.percona.com/job/percona-server-8.0-param/375/

File: storage/innobase/sync/sync0arr.cc [hunk 5]
```diff
@@ -863,9 +863,118 @@ static bool sync_array_detect_deadlock(
...
+}
+
+/** If the wakeup algorithm does not work perfectly at semaphore relases,
+ this function will do the waking (see the comment in mutex_exit). This
+ function should be called about every 1 second in the server.
+
+ Note that there's a race condition between this thread and mutex_exit
+ changing the lock_word and calling signal_object, so sometimes this finds
+ threads to wake up even when nothing has gone wrong. */
+void sync_arr_wake_threads_if_sema_free(void) {
```
Comment (percona-ysorokin) [inline storage/innobase/sync/sync0arr.cc:972]: It looks like `sync_arr_wake_threads_if_sema_free()` is not used anywhere

File: storage/innobase/include/sync0rw.ic (hunk 4 above)
Comment (satya-bodapati) [reply storage/innobase/include/sync0rw.ic:587]: Its there in rw_lock_free_func():

  /* We did an in-place new in rw_lock_create_func() */
    ut_d(lock->~rw_lock_t());
  }

File: storage/innobase/sync/sync0arr.cc (hunk 5 above)
Comment (satya-bodapati) [reply storage/innobase/sync/sync0arr.cc:972]: Good catch. Fixed

**Comment by satya-bodapati** [comment]
@percona-ysorokin Regarding the comments on rw_lock_t (sync0rw.h), it is upstream code.

If you think they won't cause merge pain/issues, I can address them

**Comment by percona-ysorokin** [comment]
@satya-bodapati If this is upstream, let's leave it as is.

**Comment by satya-bodapati** [comment]
> @satya-bodapati If this is upstream, let's leave it as is.

Ok

File: storage/innobase/include/ib0mutex.h (hunk 1 above)
Comment (satya-bodapati) [reply storage/innobase/include/ib0mutex.h:526]: left as is. Upstream code

File: storage/innobase/include/sync0rw.h (hunk 2 above)
Comment (satya-bodapati) [reply storage/innobase/include/sync0rw.h:639]: left as is. Upstream code

**Comment by satya-bodapati** [comment]
Ready for next round of review

**Comment by satya-bodapati** [comment]
Also analyzed jenkins failures.. All of them are seen in trunk (8.0.17 - trunk issue)

Specially these one (seen in trunk too):
MySQL.auth_sec.keyring_file_data_qa
MySQL.clone.local_dml_compressed_row
MySQL.clone.remote_dml_compressed_row
MySQL.encryption.innodb_encryption-page-compression
MySQL.innodb.mysqld_core_dump_without_buffer_pool_with_resizing
MySQL.innodb.percona_changed_page_bmp_requests_debug
MySQL.main.backup_safe_binlog_info
MySQL.main.mysqldumpslow

**Comment by satya-bodapati** [comment]
New jenkins link: https://ps80.cd.percona.com/job/percona-server-8.0-param/381/

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Pull request merged**